### PR TITLE
8003417: WeakHashMap$HashIterator removes wrong entry

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review for this change which proposes to fix the issue reported in https://bugs.openjdk.java.net/browse/JDK-8003417?

The issue notes that this is applicable for `WeakHashMap` which have `null` keys. However, the issue is even applicable for `WeakHashMap` instances which don't have `null` keys, as reproduced and shown by the newly added jtreg test case in this PR.

The root cause of the issue is that once the iterator is used to iterate till the end and the `remove()` is called, then the `WeakHashMap$HashIterator#remove()` implementation used to pass `null` as the key to remove from the map, instead of the key of the last returned entry. The commit in this PR fixes that part.

A new jtreg test has been added which reproduces the issue as well as verifies the fix.
`tier1` testing and this new test have passed after this change. However, I guess this will require a JCK run to be run too, perhaps? If so, I will need help from someone who has access to them to have this run against those please.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8003417](https://bugs.openjdk.java.net/browse/JDK-8003417): WeakHashMap$HashIterator removes wrong entry

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6488/head:pull/6488` \
`$ git checkout pull/6488`

Update a local copy of the PR: \
`$ git checkout pull/6488` \
`$ git pull https://git.openjdk.java.net/jdk pull/6488/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6488`

View PR using the GUI difftool: \
`$ git pr show -t 6488`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6488.diff">https://git.openjdk.java.net/jdk/pull/6488.diff</a>

</details>

### CODE REVIEW

**Comment by jaikiran** [comment]
Any reviews for this change, please?

**Comment by RogerRiggs** [comment]
fyi, with this change the JCK did not flag a failure.

**Comment by jaikiran** [comment]
> fyi, with this change the JCK did not flag a failure.

Thank you Roger for running those tests.

**Review by RogerRiggs — APPROVED** [review]
Looks good to me.  Another set of eyes would be good on this fringe case.

File: src/java.base/share/classes/java/util/WeakHashMap.java [hunk 1]
```diff
@@ -823,7 +823,7 @@ public void remove() {
             if (modCount != expectedModCount)
                 throw new ConcurrentModificationException();
 
-            WeakHashMap.this.remove(currentKey);
+            WeakHashMap.this.remove(lastReturned.get());
```
Comment (bchristi-git) [inline src/java.base/share/classes/java/util/WeakHashMap.java:826]: Unlike `currentKey`, `HashIterator` does not necessarily maintain a strong reference to the referent of `lastReturned`. If the `lastReturned` WeakReference were cleared between `lastReturned` being set, and the `remove()` call, the null key could again be erroneously removed. It would be cool to add a test case for this (maybe using large objects as keys would tempt the GC to clear the weakrefs via `jdk.test.lib.util.ForceGC`).

Since we store the `NULL_KEY` sentinel value for the null key, I believe `Entry.get()` will only return null in the case that the weakref has been cleared.  So could we instead fix this with:

```
if (lastReturned.get() != null) {
    WeakHashMap.this.remove(lastReturned.get());
}
```

?
Or, if we're worried about the ref being cleared between the null check and remove(), maybe:
```
Object lastReturnedKey = lastReturned.get();
if (lastReturnedKey != null) {
    WeakHashMap.this.remove(lastReturned.get());
}
```

**Review by bchristi-git — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/util/WeakHashMap.java (hunk 1 above)
Comment (stuart-marks) [reply src/java.base/share/classes/java/util/WeakHashMap.java:826]: Yes, this is my concern too; lastReturned is a weak ref, and it is possible for it to have been cleared at this point, so get() can return null. I will post a more complete explanation.

**Comment by stuart-marks** [comment]
I've put extensive comments and explanation into the bug report. I put them there instead of here because they're more closely related to the test case in that bug report, and they also include general statements about WeakHashMap iteration, not specific to the change proposed here.

TL;DR I don't think that changing remove to `remove(lastReturned.get())` is the correct fix. It requires revisting the invariants maintained by `HashIterator`. It might be as simple as removing the clearing of `currentKey` from `hasNext()`, as mentioned in the bug report, but this requires more thought.

**Review by RogerRiggs — CHANGES_REQUESTED** [review]
The value of `currentKey` should track the value of `lastReturned.get()` and does so in the `nextEntry()` and `remove()` methods.  
The odd statement is in `hasNext` that set the state of `currentkey` to null independently of `lastReturned`.
Leaving it non-null will allow `remove` to operate correctly but may have a downside of keeping a strong reference to last key until the HashIterator is GC'd.
The `remove()` method can correctly use `currentKey` to remove the last entry returned.

**Comment by stuart-marks** [comment]
/open

Seems like we ought to be able to fix this one.

**Comment by jaikiran** [comment]
> Seems like we ought to be able to fix this one.

**Comment by jaikiran** [comment]
> Seems like we ought to be able to fix this one.

Agreed. I got caught up in a few unrelated things, so haven't been able to get back to this to take into account some of your and Roger's inputs. I'll get back to this shortly.

**Comment by jaikiran** [comment]
(work is currently in progress on this one)
